File: gameMap.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <iterator> 
#include <vector>
#include "engimon.hpp"
#include "tile.hpp"
#include "gameMap.hpp"
#include <stdio.h>      
#include <stdlib.h>
#include <time.h>


gameMap::gameMap()
{
	tile_map = new tile*[this->mapWidth];
	for (int k = 0; k < this->mapWidth; ++k)
	{
		tile_map[k] = new tile[this->mapLength];
	}


	engimon nullEngimon;
	for (int i = 0; i < this->mapWidth; ++i)
	{
		for (int j = 0; j < this->mapLength	; ++j)
		{
			if (i < this->mapLength/2 && j > this->mapWidth/2)
			{
				tile_map[i][j] = tile(i,j,nullEngimon,"sea");
			}
			else
			{
				tile_map[i][j] = tile(i,j,nullEngimon,"grassland");
			}
			
		}
	}
}

gameMap::gameMap(string externalFile)
{
	ifstream f(externalFile);

	if (!f.is_open())
	{
		throw "File tidak dibisa dibuka atau tidak ada";
	}
	int charCount = 0;
	int baris, kolom;
	char c;
	f >> baris;
	f >> kolom;

	while (f >> c)
	{
		charCount++;
	}

	if (charCount != baris*kolom)
	{
		throw "Input file salah, periksa ulang";
	}

	f = ifstream(externalFile);
	f >> baris;
	f >> kolom;
	this->mapLength = kolom;
	this->mapWidth = baris;

	tile_map = new tile*[this->mapWidth];
	for (int k = 0; k < this->mapWidth; ++k)
	{
		tile_map[k] = new tile[this->mapLength];
	}


	engimon nullEngimon;
	for (int i = 0; i < this->mapWidth; ++i)
	{
		for (int j = 0; j < this->mapLength	; ++j)
		{
			f >> c;
			if (c == 'o')
			{
				tile_map[i][j] = tile(i,j,nullEngimon,"sea");
			}
			else
			{
				tile_map[i][j] = tile(i,j,nullEngimon,"grassland");
			}
			
		}
	}
}

gameMap::~gameMap()
{
	for (int i = 0; i < this->mapWidth; ++i)
	{
		delete[] tile_map[i];
	}
	delete[] tile_map;
}

void gameMap::printMap()
{
	for (int i = 0; i < this->mapWidth; ++i)
	{
		for (int j = 0; j < this->mapLength	; ++j)
		{
			tile_map[i][j].printTile();
			cout << " ";
		}
		cout << endl;
	}
}
void gameMap::printLegend()
{
	cout << "Legend:" << endl;
	cout << "W/w: Water engimon" << endl;
	cout << "I/i : Ice engimon" << endl;
	cout << "F/f: Fire engimon" << endl;
	cout << "G/g: Ground engimon" << endl;
	cout << "E/e: Electric engimon" << endl;
	cout << "L/l: Fire/Electric engimon" << endl;
	cout << "S/s: Water/Ice engimon" << endl;
	cout << "N/n: Water/Ground engimon" << endl;

	cout << endl;
	cout << "P: Player" << endl;
	cout << "X: Active Engimon" << endl;

	cout << endl;
	cout << "- : Grassland" << endl;
	cout << "o : Sea" << endl;
}

void moveTileEngimon(tile &tile1, tile &tile2)
{
	/*Menggerakan pokemon dari tile1 ke tile2*/

	if (tile2.haveWildEngimon() == false && tile2.isTileValid(tile1.getEngimon()) && tile2.isPlayerHere() == false && tile2.isactiveEngimonHere() == false)
	{
		engimon nullEngimon;

		tile2.setEngimon(tile1.getEngimon());
		tile1.setEngimon(nullEngimon);
	}

}

void gameMap::moveWildEngimon()
{
	/*Random Module*/
	srand(tile::random_seed);
	tile::random_seed++;
	int randomDirection;
	randomDirection = rand() %  4;
	srand(tile::random_seed);
	tile::random_seed++;

	int i, j;

	/*First pass, find all empty tiles*/
	for (i = 0; i < this->mapWidth; ++i)
	{
		for (j = 0; j < this->mapLength; ++j)
		{
			if (tile_map[i][j].haveWildEngimon() == false)
			{
				tile_map[i][j].passed();
			}
		}
	}

	/*Second pass, move semua tile yang ada pokemon*/
	for (i = 0; i < this->mapWidth; ++i)
	{
		for (j = 0; j < this->mapLength	; ++j)
		{
			if (tile_map[i][j].haveWildEngimon() && tile_map[i][j].isPass() == false)
			{
				// cout << j << "," << i << " ";

				// tile_map[i][j].getEngimon().showAttributes();
				// cout << endl;
				
				if (randomDirection == 0 && i > 0) //up
				{
					moveTileEngimon(tile_map[i][j], tile_map[i-1][j]);
					// cout << "move up" << endl;
				}
				else if (randomDirection == 1 && j > 0)//left
				{
					moveTileEngimon(tile_map[i][j], tile_map[i][j-1]);
					// cout << "move left" << endl;
				}
				else if (randomDirection == 2 && j < this->mapLength - 1)//right
				{
					moveTileEngimon(tile_map[i][j], tile_map[i][j+1]);
					// cout << "move right" << endl;
				}
				else
				{
					if (randomDirection == 3 && i < this->mapWidth - 1)//down
					{
						moveTileEngimon(tile_map[i][j], tile_map[i+1][j]);
						// cout << "move down" << endl;
					}
				}
				tile_map[i][j].passed();
				randomDirection = rand() %  4;
				srand(time(NULL));
			}
		}
	}

	for (i = 0; i < this->mapWidth; ++i)
	{
		for (j = 0; j < this->mapLength; ++j)
		{
			tile_map[i][j].resetPass();
		}
	}

}
void gameMap::generateEngimon()
{
	/*Random Module*/
	int seed = time(NULL);
	while(engimon_count < MAP_ENGIMON_COUNT)
	{
		srand(seed);
		int randomX, randomY;
		randomX = rand() %  this->mapLength;
		srand(seed);
		randomY = rand() % this->mapWidth;

		//cout << randomX << "," << randomY << endl;

		if (tile_map[randomY][randomX].haveWildEngimon() == false)
		{
			tile_map[randomY][randomX].spawn();
			engimon_count++;
		}
		seed++;
	}
}
	
void gameMap::updateMap(int player_x, int player_y, int active_x, int active_y)
{
	/*y itu baris, x itu column*/
	for (int i = 0; i < this->mapWidth; ++i)
	{
		for (int j = 0; j < this->mapLength; ++j)
		{
			tile_map[i][j].playerIsNotHere();
			tile_map[i][j].activeEngimonIsNotHere();
		}
	}

	if (tile_map[active_y][active_x].haveWildEngimon())
	{
		throw "Active Engimon menabrak Wild Engimon";
	}

	this->tile_map[player_y][player_x].playerIsHere();
	this->tile_map[active_y][active_x].activeEngimonIsHere();
	map_move_count--;

	if (map_move_count <= 0)
	{
		moveWildEngimon();
		map_move_count = 5;
	}
	this->tile_map[player_y][player_x].playerIsHere();
	this->tile_map[active_y][active_x].activeEngimonIsHere();
}

void gameMap::deleteTileEngimon(int x, int y)
{
	engimon nullEngimon;
	if (!(tile_map[y][x].isPlayerHere() || tile_map[y][x].isactiveEngimonHere()) && tile_map[y][x].haveWildEngimon())
	{
		this->tile_map[y][x].setEngimon(nullEngimon);
		engimon_count--;
		if (engimon_count < 3)
		{
			generateEngimon();
		}
	}
	
}


engimon& gameMap::getTileEngimon(int x, int y)
{
	return tile_map[y][x].getEngimon();
}
string gameMap::findNearbyEngimon(int x, int y)
{
	if (y > 0 && tile_map[y-1][x].haveWildEngimon())
	{
		return "up";
	}
	else if (x > 0 && tile_map[y][x-1].haveWildEngimon())
	{
		return "left";
	}
	else if (y < this->mapWidth-1 &&tile_map[y+1][x].haveWildEngimon())
	{
		return "up";
	}
	else if (x < this->mapLength-1 && tile_map[y][x+1].haveWildEngimon())
	{
		return "right";
	}
	else if (tile_map[y][x].haveWildEngimon())
	{
		return "center";
	}
	else
	{
		return "null";
	}
}

bool gameMap::isTileOccupied(int x, int y)
{
	return (tile_map[y][x].haveWildEngimon());
}

int gameMap::getMapLength()
{
	return mapLength;
}
int gameMap::getMapWidth()
{
	return mapWidth;
}

File: Java/gameMap.java
import java.io.File;
import java.util.Random;
import java.util.Scanner;
import java.util.Vector;

public class gameMap {
    /*Gunakan Vector agar thread safe*/
    protected Vector<Vector<tile>> tile_map;

    private int mapLength = 12;
    private int mapWidth = 10;

    protected int engimon_count = 0;
    private int map_move_count = MAP_MOVE;
    private int map_generate_count = MAP_GENERATE;
    private int map_wild_engimon_levelup = MAP_WILD_ENGIMON_LEVELUP;

    public static final int MAP_ENGIMON_COUNT = 5;
    public static final int MAP_MOVE = 5;
    public static final int MAP_GENERATE = 10;
    public static final int MAP_WILD_ENGIMON_LEVELUP = 30;

    public gameMap()
    {
        tile_map = new Vector<Vector<tile>>();
        for (int i = 0; i < this.mapWidth; i++) {
            Vector<tile> r = new Vector<tile>();
            for (int j = 0; j < this.mapLength; j++) {
                if (i < this.mapLength/2 && j < this.mapWidth/2)
                {
                    r.add(new tile(j,i,"sea"));
                }
			    else if (i >= this.mapLength/2 && j < this.mapWidth/2)
                {
                    r.add(new tile(j,i,"grassland"));
                }
			    else if (i < this.mapLength/2 && j >= this.mapWidth/2)
                {
                    r.add(new tile(j,i,"mountains"));
                }
			    else
                {
                    r.add(new tile(j,i,"tundra"));
                }
            }
            tile_map.add(r);
        }
    }
    public gameMap(String externalFile) throws Exception
    {
        File f;
        Scanner fileReader;
        try
        {
            f = new File(externalFile);
            fileReader = new Scanner(f);

        }
        catch (Exception e)
        {
            throw e;
        }

        String line = fileReader.nextLine();
        String[] firstLine= line.split(" ");
        Integer baris = Integer.parseInt(firstLine[0]);
        Integer kolom = Integer.parseInt(firstLine[1]);

        Integer count = 0;
        while (fileReader.hasNextLine())
        {
            line = fileReader.nextLine();
            count++;
            if (line.split(" ").length != kolom)
            {
                throw new Exception("Pastikan input file benar");
            }
        }

        if (count != baris)
        {
            throw new Exception("Pastikan input file benar");
        }
        fileReader.close();

        fileReader = new Scanner(f);
        line = fileReader.nextLine();
        this.mapWidth = baris;
        this.mapLength = kolom;
        tile_map = new Vector<Vector<tile>>();
        for (int i = 0; i < this.mapWidth; i++) {
            line = fileReader.nextLine();
            String[] lineArray = line.split(" ");
            Vector<tile> r = new Vector<tile>();
            for (int j = 0; j < this.mapLength; j++) {
                if ("o".equals(lineArray[j]))
                {
                    r.add(new tile(j,i,"sea"));
                }
                else if (".".equals(lineArray[j]))
                {
                    r.add(new tile(j,i,"tundra"));
                }
                else if ("^".equals(lineArray[j]))
                {
                    r.add(new tile(j,i,"mountains"));
                }
                else
                {
                    r.add(new tile(j,i,"grassland"));
                }
            }
            tile_map.add(r);
        }
        fileReader.close();
    }

    public void printMap()
    {
        for (Vector<tile> i :tile_map){
            for (tile j:i) {
                j.printTile();
                System.out.print(" ");
            }
            System.out.println("");
        }
    }

    public void printLegend()
    {
        System.out.println("Legend:");
        System.out.println("W/w: Water engimon");
        System.out.println("I/i : Ice engimon");
        System.out.println("F/f: Fire engimon");
        System.out.println("G/g: Ground engimon");
        System.out.println("E/e: Electric engimon");
        System.out.println("L/l: Fire/Electric engimon");
        System.out.println("S/s: Water/Ice engimon");
        System.out.println("N/n: Water/Ground engimon");

        System.out.println();
        System.out.println("P: Player");
        System.out.println("X: Active Engimon");

        System.out.println();
        System.out.println("- : Grassland");
        System.out.println("o : Sea");
        System.out.println("^ : Mountains>");
        System.out.println(". : Tundra");
    }

    private int randomInteger(int start, int end)
    {
        Random rand = new Random();

        int rand_int = rand.nextInt(end-start+1);
        return rand_int + start;
    }

    public void moveTileEngimon(tile tile1, tile tile2)
    {
        /*Menggerakan pokemon dari tile1 ke tile2*/
        if (tile2.haveWildEngimon() == false && tile2.isTileValid(tile1.getEngimon()) && tile2.isPlayerHere() == false && tile2.isactiveEngimonHere() == false)
        {
            engimon nullEngimon;

            tile2.setEngimon(tile1.getEngimon());
            tile1.setEngimon(null);
        }
    }

    public void moveWildEngimon()
    {
        for (Vector<tile> i :tile_map){
            for (tile j:i) {
                if (j.haveWildEngimon() == false)
                {
                    j.passed();
                }
            }
        }

        Integer randomDirection = (randomInteger(0,10000) % 4);

        for (Vector<tile> i :tile_map){
            for (tile j:i) {
                if (j.haveWildEngimon() && j.isPass() == false)
                {
                    if (randomDirection == 0 && j.y > 0) //up
                    {
                        moveTileEngimon(j, tile_map.elementAt(j.y-1).elementAt(j.x));
                        // cout << "move up" << endl;
                    }
                    else if (randomDirection == 1 && j.x > 0)//left
                    {
                        moveTileEngimon(j, tile_map.elementAt(j.y).elementAt(j.x - 1));
                        // cout << "move left" << endl;
                    }
                    else if (randomDirection == 2 && j.x < this.mapLength - 1)//right
                    {
                        moveTileEngimon(j, tile_map.elementAt(j.y).elementAt(j.x+1));
                        // cout << "move right" << endl;
                    }
				    else
                    {
                        if (randomDirection == 3 && j.y < this.mapWidth - 1)//down
                        {
//                            System.out.println(randomDirection);
//                            assert (randomDirection == 3);
                            moveTileEngimon(j, tile_map.elementAt(j.y+1).elementAt(j.x));
                            // cout << "move down" << endl;
                        }
                    }
                    j.passed();
                    randomDirection = (randomInteger(0,10000) % 4);
                }
            }
        }

        for (Vector<tile> i :tile_map){
            for (tile j:i) {
                j.resetPass();

            }
        }
    }

    public void generateEngimon()
    {
        while (engimon_count < MAP_ENGIMON_COUNT)
        {
            int randomX = randomInteger(0,this.mapLength - 1);
            int randomY = randomInteger(0,this.mapWidth - 1);

            if (this.tile_map.elementAt(randomY).elementAt(randomX).haveWildEngimon() == false)
            {
                this.tile_map.elementAt(randomY).elementAt(randomX).spawn();
                this.engimon_count++;
            }
        }
    }

    public void levelUpEngimon()
    {
        for (Vector<tile> i :tile_map){
            for (tile j:i) {
                if (j.haveWildEngimon())
                {
                    j.getEngimon().gainExp(100);
                    if (j.getEngimon().getCummulativeExp() >= 10000)
                    {
                        deleteTileEngimon(j.x,j.y);
                    }
                }
            }
        }
    }


    public void updateMap(int player_x, int player_y, int active_x, int active_y) throws Exception
    {
        for (Vector<tile> i :tile_map){
            for (tile j:i) {
                j.playerIsNotHere();
                j.activeEngimonIsNotHere();
            }
        }
        if (this.tile_map.elementAt(active_y).elementAt(active_x).haveWildEngimon())
        {
            throw new Exception("Active Engimon menabrak Wild Engimon");
        }

        this.tile_map.elementAt(player_y).elementAt(player_x).playerIsHere();
        this.tile_map.elementAt(active_y).elementAt(active_x).activeEngimonIsHere();

        this.map_move_count = this.map_move_count - 1;
        if (this.map_move_count <= 0)
        {
            moveWildEngimon();
            map_move_count = MAP_MOVE;
        }

        this.map_generate_count = this.map_generate_count - 1;
        if (this.map_generate_count <= 0)
        {
            generateEngimon();
            map_generate_count = MAP_GENERATE;
        }

        this.map_wild_engimon_levelup = this.map_wild_engimon_levelup - 1;
        if (this.map_wild_engimon_levelup <= 0)
        {
            levelUpEngimon();
            map_wild_engimon_levelup = MAP_WILD_ENGIMON_LEVELUP;
        }

        this.tile_map.elementAt(player_y).elementAt(player_x).playerIsHere();
        this.tile_map.elementAt(active_y).elementAt(active_x).activeEngimonIsHere();
    }

    public void updateMap(int player_x, int player_y, int active_x, int active_y, player P) throws Exception
    {
        this.updateMap(player_x, player_y, active_x, active_y);
        this.tile_map.elementAt(active_y).elementAt(active_x).activeEngimonIsHere(P.getActiveEngimon());
    }

    public void deleteTileEngimon(int x, int y)
    {
        if (!(tile_map.elementAt(y).elementAt(x).isPlayerHere() || tile_map.elementAt(y).elementAt(x).isactiveEngimonHere())
                && tile_map.elementAt(y).elementAt(x).haveWildEngimon())
        {
            this.tile_map.elementAt(y).elementAt(x).setEngimon(null);
            engimon_count = engimon_count - 1;
            if (engimon_count < 3)
            {
                generateEngimon();
            }
        }
    }
    public engimon getTileEngimon(int x, int y)
    {
        return tile_map.elementAt(y).elementAt(x).getEngimon();
    }
    public String findNearbyEngimon(int x, int y)
    {
        if (y > 0 && tile_map.elementAt(y-1).elementAt(x).haveWildEngimon())
        {
            return "up";
        }
        else if (x > 0 && tile_map.elementAt(y).elementAt(x-1).haveWildEngimon())
        {
            return "left";
        }
        else if (y < this.mapWidth-1 &&tile_map.elementAt(y+1).elementAt(x).haveWildEngimon())
        {
            return "up";
        }
        else if (x < this.mapLength-1 && tile_map.elementAt(y).elementAt(x+1).haveWildEngimon())
        {
            return "right";
        }
        else if (tile_map.elementAt(y).elementAt(x).haveWildEngimon())
        {
            return "center";
        }
        else
        {
            return "null";
        }

    }
    public Boolean isTileOccupied(int x, int y)
    {
        return (tile_map.elementAt(y).elementAt(x).haveWildEngimon() && !tile_map.elementAt(y).elementAt(x).getEngimon().isActive());
    }

    public int getMapLength(){ return this.mapLength; }
    public int getMapWidth(){ return this.mapWidth; }
}
